[compression-dictionary] Allow for larger ZStandard compression windows. (#2780)
httpwg/http-extensions: draft-ietf-httpbis-compression-dictionary.md

Content-Dictionary: :pZGm1Av0IEBKARczz7exkNYsZb8LzaMrV7J32a2fFG4=:
~~~

# Dictionary-Compressed Brotli

The "br-d" content encoding identifies a resource that is a
"Dictionary-Compressed Brotli" stream.

A "Dictionary-Compressed Brotli" stream is a Brotli {{RFC7932}} stream for a
response that has been compressed with an external dictionary using a
compression window not larger than 16 MB.

Clients that announce support for br-d content encoding MUST be able to
decompress resources that were compressed with a window size of up to 16 MB.

With Brotli compression, the full dictionary is available during compression
and decompression independent of the compression window, allowing for
delta-compression of resources larger than the compression window.

# Dictionary-Compressed Zstandard

The "zstd-d" content encoding identifies a resource that is a
"Dictionary-Compressed Zstandard" stream.

A "Dictionary-Compressed Zstandard" stream is a Zstandard {{RFC8878}} stream

> for a response that has been compressed with an external dictionary using a
> compression window not larger than the larger of 8 MB and 1.25 times the size
> of the dictionary, not to exceed 128 MB.

Reviewer: ```suggestion
for a response that has been compressed with an external dictionary.
```

Clients that announce support for zstd-d content encoding MUST be able to
decompress resources that were compressed with a window size of up to the

> larger of 8 MB and 1.25 times the size of the dictionary, not to exceed 128 MB.

Reviewer: Do you really need to repeat this?  I'd keep this and drop the earlier version.

> decompress resources that were compressed with a window size of up to the
> larger of 8 MB and 1.25 times the size of the dictionary, not to exceed 128 MB.

Reviewer: ```suggestion
decompress resources that were compressed with a window size of at least 8 MB
or 1.25 times the size of the dictionary, which ever is greater, up to a maximum of 128 MB.
```

Reviewer: Or, with a bonus text.

```suggestion
decompress resources that were compressed with a window size of at least 8 MB
or 1.25 times the size of the dictionary, which ever is greater, up to a maximum of 128 MB.
Clients that are not willing to accept a larger window need to avoid advertising
any dictionary that would require a larger window.
```

What needs to be said about rounding?  12345 * 1.25 = what?  Are we rounding up or down, to a power of two?

Author: On the bonus language, I think it is a bit more subtle because they could advertise a dictionary but only `br-d` if they want to keep the window under 16 MB (and if they support both).

As far as rounding goes, as far as I understand, the window representation in the stream is where the power of two (rounding down) happens so that part of it isn't something we need to spec since accepting the larger non-rounded window is still acceptable, there is just no way to represent it in the stream.

That said, it's worth mentioning rounding on the actual 1.25 times calculation to avoid fractional bytes. I added some language to round it down (effectively what you would get with integer math on most languages).

The "br-d" content encoding identifies a resource that is a
"Dictionary-Compressed Brotli" stream.

A "Dictionary-Compressed Brotli" stream is a Brotli {{RFC7932}} stream for a
response that has been compressed with an external dictionary using a
compression window not larger than 16 MB.

Clients that announce support for br-d content encoding MUST be able to
decompress resources that were compressed with a window size of up to 16 MB.

With Brotli compression, the full dictionary is available during compression
and decompression independent of the compression window, allowing for
delta-compression of resources larger than the compression window.

# Dictionary-Compressed Zstandard

The "zstd-d" content encoding identifies a resource that is a
"Dictionary-Compressed Zstandard" stream.

A "Dictionary-Compressed Zstandard" stream is a Zstandard {{RFC8878}} stream
for a response that has been compressed with an external dictionary.

Clients that announce support for zstd-d content encoding MUST be able to
decompress resources that were compressed with a window size of at least 8 MB
or 1.25 times the size of the dictionary, which ever is greater, up to a
maximum of 128 MB.

> If 1.25 times the size of the dictionary results in a fractional byte window
> size, the window size should be rounded down to the nearest byte.

Reviewer: Given that the stream will include a window size as a power of two, maybe you should just say:

```suggestion
The window size used will be encoded in the content (currently, this can be expressed
in powers of two only) and it MUST be lower than this limit. An implementation MAY
treat a window size that exceeds the limit as a decoding error.
```

## Compression Algorithms
This document introduces two new content encoding algorithms:

- br-d: Brotli {{RFC7932}} using an external compression dictionary.
- zstd-d: Zstandard {{RFC8878}} using an external compression dictionary.

The client and server need to agree on a maximum size to allow for the
compression window before a request is made so that the server can be sure that
the client will be able to decompress the response.

Dictionary-based compression benefits from having the full dictionary available
whenever possible, particularly for cases where a previous version of a
resource is being used as a dictionary for the newer version (delta
compression).

### Brotli Compression Window

Brotli {{RFC7932}} makes the full dictionary available independent of the
compression window so the br-d content encoding allows for the same 16 MB
maximum compression window as the br content encoding.

Clients that announce support for br-d content encoding MUST be able to

> decompress resources that were compressed with a window size of up to 16 MB.

Reviewer 2: ```suggestion
decompress resources that were compressed with a window size of not more than 16 MB.
```

> ### ZStandard Compression Window

Reviewer 3: nit: here and below, the capitalization is "Zstandard" rather than "ZStandard".

Zstandard {{RFC8878}} makes the full dictionary available as long as the input
is smaller than the compression window. Once the input grows beyond the
compression window the dictionary becomes unavailable.

In order to allow for large resources to be dictionary-compressed while still
giving clients control over how much memory they will consume, the zstd-d

> content encoding allows for a compression window of up to the larger of 8 MB

Reviewer 2: ```suggestion
content encoding allows for a compression window of not more than the larger of 8 MB
```

and the size of the dictionary.

For dictionaries that exceed the amount of memory that a client is willing to
allocate for decompression, the client can choose to not announce support for
zstd-d content encoding.

> Clients that announce support for br-d content encoding MUST be able to

Reviewer 3: nit: `br-d` -> `zstd-d`.

> decompress resources that were compressed with a window size of up to the

Reviewer 2: ```suggestion
decompress resources that were compressed with a window size of not more than the
```

## Content Encoding

IANA is asked to enter the following into the "HTTP Content Coding Registry"
registry ({{HTTP}}):

- Name: br-d
- Description: A stream of bytes compressed using the Brotli protocol with an

> external dictionary of not more than 16 MB.

Author: Whoops. yes, thanks. Copy/paste problem where I accidentally cropped out too much of the description last time.

> external dictionary and a compression window of not more than 16 MB.

Reviewer: Now that the window size is more complex, I would drop this additional text.  Instead, reference the definition of the window size above.  (I would structure this differently and have sections for each algorithm, that includes a discussion of window size, rather than sections for the window size.)

Reviewer 2: It seems nice to keep the requirement itself in the IANA registry, and have the longer discussion and reasoning up above.

Author: I changed it to be a named format instead of trying to include all of the requirements and settings directly in the IANA registry.  They're now referenced as `Dictionary-Compressed Brotli` and `Dictionary-Compressed Zstandard`

- Reference: This document
- Notes: {{compression-algorithms}}

IANA is asked to enter the following into the "HTTP Content Coding Registry"
registry ({{HTTP}}):

- Name: zstd-d
- Description: A stream of bytes compressed using the Zstandard protocol with
an external dictionary of not more than the larger of 8 MB and the size of the

> dictionary.

Reviewer 3: If we want to keep the discussion of the window size here, I would propose phrasing this as:

> A stream of bytes compressed using the Zstandard protocol using a negotiated external dictionary. Unlike the `zstd` Content Coding, the Zstandard frame(s) making up the stream may use a `Window_Size` up to and including the size of the external dictionary or 8 MB, whichever is larger.

This makes it clear that the constraint is on the window, and not the size of the dictionary: "an external dictionary of not more than [...] the size of the dictionary" is tautological.

If we take @Cyan4973's suggestion in [this comment](https://github.com/httpwg/http-extensions/issues/2754#issuecomment-2061998009) (and I agree wiggle room would be nice to have here) I guess we'll have to complicate this a little further.

But since you are actually specifying this above, maybe it doesn't even need to be mentioned here.

Author: The main reason to mention it here is that this chunk of text is what is going to be in the IANA registry for the content encodings.  I fixed the language around it being the window size and not the dictionary, thanks.
